packages/nimble-components/src/table/models/keyboard-navigation-manager.ts

```diff
@@ -0,0 +1,1289 @@
+import { Notifier, Subscriber, Observable } from '@microsoft/fast-element';
+import {
+    keyArrowDown,
+    keyArrowLeft,
+    keyArrowRight,
+    keyArrowUp,
+    keyEnd,
+    keyEnter,
+    keyEscape,
+    keyFunction2,
+    keyHome,
+    keyPageDown,
+    keyPageUp,
+    keySpace,
+    keyTab
+} from '@microsoft/fast-web-utilities';
+import { FoundationElement } from '@microsoft/fast-foundation';
+import type { ScrollToOptions } from '@tanstack/virtual-core';
+import type { Table } from '..';
+import {
+    TableActionMenuToggleEventDetail,
+    TableFocusType,
+    TableHeaderFocusableElements,
+    TableRowFocusableElements,
+    type TableRecord
+} from '../types';
+import type { Virtualizer } from './virtualizer';
+import { TableGroupRow } from '../components/group-row';
+import { TableRow } from '../components/row';
+import { TableCell } from '../components/cell';
+import { MenuButton } from '../../menu-button';
+import { tableHeaderTag } from '../components/header';
+import { TableCellView } from '../../table-column/base/cell-view';
+
+interface TableFocusState {
+    focusType: TableFocusType;
+    headerActionIndex?: number;
+    rowIndex?: number;
+    columnIndex?: number;
+    cellContentIndex?: number;
+}
+
+/**
+ * Manages the keyboard navigation and focus within the table.
+ * @internal
+ */
+export class KeyboardNavigationManager<TData extends TableRecord>
+implements Subscriber {
+    private focusType: TableFocusType = TableFocusType.none;
+    private headerActionIndex = -1;
+    private rowIndex = -1;
+    private cellContentIndex = -1;
+    private columnIndex = -1;
+    private readonly tableNotifier: Notifier;
+    private readonly virtualizerNotifier: Notifier;
+    private visibleRowNotifiers: Notifier[] = [];
+    private get inNavigationMode(): boolean {
+        return (
+            this.focusType !== TableFocusType.cellActionMenu
+            && this.focusType !== TableFocusType.cellContent
+        );
+    }
+
+    public constructor(
+        private readonly table: Table<TData>,
+        private readonly virtualizer: Virtualizer<TData>
+    ) {
+        table.addEventListener('keydown', e => this.onCaptureKeyDown(e), {
+            capture: true
+        });
+        table.addEventListener('keydown', e => this.onKeyDown(e));
+        table.addEventListener('focusin', e => this.handleFocus(e));
+        this.tableNotifier = Observable.getNotifier(this.table);
+        this.tableNotifier.subscribe(this, 'rowElements');
+        this.virtualizerNotifier = Observable.getNotifier(this.virtualizer);
+        this.virtualizerNotifier.subscribe(this, 'visibleItems');
+    }
+
+    public resetFocusState(): void {
+        this.focusType = TableFocusType.none;
+        const activeElement = this.getActiveElement();
+        if (activeElement && this.isInTable(activeElement)) {
+            this.setDefaultFocus();
+        }
+    }
+
+    public connect(): void {
```

Comment:
I think we're missing a `disconnect()` method that gets called from the table's `disconnectedCallback()`. Right now it looks like we'd get duplicate event listeners if you create a table and add/remove/add it to the DOM.

Reply:
Added `disconnect()`.